### ISSUE TITLE
Caused by: STATUS_SHARING_VIOLATION(3221225539/3221225539): Create failed for Harisingh\\\DBCompareReport\Output\\9YJUPeB6.xml

### ISSUE THREAD

**Issue by harisingh-highq** [hierynomus/smbj#356]
Hi All
**We found below exception for different file operations in our live environment:-**
`Caused by: STATUS_SHARING_VIOLATION(3221225539/3221225539): Create failed for Harisingh\\\DBCompareReport\Output\\9YJUPeB6.xml
	at com.hierynomus.smbj.share.Share.receive(Share.java:342)
	at com.hierynomus.smbj.share.Share.sendReceive(Share.java:322)
	at com.hierynomus.smbj.share.Share.createFile(Share.java:135)
	at com.hierynomus.smbj.share.DiskShare.createFileAndResolve(DiskShare.java:79)
	at com.hierynomus.smbj.share.DiskShare.open(DiskShare.java:69)
	at com.hierynomus.smbj.share.DiskShare.openFile(DiskShare.java:144)
	at com.hybridintegration.repository.SMBJRepo.writeOutputXmlToRepository(SMBJRepo.java:2382)
	... 7 more`

**As i was able to regenerate it locally and found one of the cause is in  below code:-**
`		Calendar cal = Calendar.getInstance();
		String filePath = null;
		logger.debug("SMBJ - In Deleting older report files ");
		try {
			if (repConnString.contains(DBCompareAPIConstants.SMBJ_INPUT_FOLDER)
					|| repConnString.contains(DBCompareAPIConstants.SMBJ_OUTPUT_FOLDER)) {
				repConnString = checkFolderPath(repConnString);
				if (share.folderExists(repConnString)) {
					for (FileIdBothDirectoryInformation f : share.list(repConnString, "*.xml")) {
						com.hierynomus.smbj.share.File smbFileOpen = null;
						try {
							filePath = repConnString + f.getFileName();
							smbFileOpen = share.openFile(filePath, EnumSet.of(AccessMask.GENERIC_READ), null,
									SMB2ShareAccess.ALL, SMB2CreateDisposition.FILE_OPEN, null);
							FileTime fileModifiedTime = smbFileOpen.getFileInformation().getBasicInformation()
									.getLastWriteTime();
							long diff = cal.getTimeInMillis() - fileModifiedTime.toEpochMillis();
							long diffHours = diff / (60 * 60 * 1000);
							if (diffHours > DBCompareAPIConstants.DELETE_FILE_OLDER_HOURS) {
								logger.info("SMBJ - DBCompare report file deleted : " + f.getFileName());
								share.rm(filePath);
							}
						} catch (Exception e) {
							logger.error("SMBJ - Exception in deleteOlderReports");
							logger.error("SMBJ - " + ExceptionUtilityHelper.accessExceptionStackTrace(e));
						} finally {
							try {
								if (smbFileOpen != null) {
									smbFileOpen.close();
								}
							} catch (Exception e) {
								logger.error("SMBJ - Exception in closing smb file object");
								logger.error("SMBJ - " + ExceptionUtilityHelper.accessExceptionStackTrace(e));
							}
						}
					}
				}
			}
		} catch (Exception e) {
			logger.error("SMBJ - Exception in deleteOlderReports");
			logger.error("SMBJ - " + ExceptionUtilityHelper.accessExceptionStackTrace(e));
		}
	String outputPath = sharedDirectory+outputFolderPath+token+".xml";
				smbJFile = share.openFile(outputPath, EnumSet.of(AccessMask.GENERIC_WRITE, 
                                AccessMask.GENERIC_READ), null, null,
				SMB2CreateDisposition.FILE_OVERWRITE_IF, null);`

Cause:- As you can see in above code, we iterates list of file under a folder and in iterating, we open every **smbFileOpen** file object by share.openFile lines of code. But close **smbFileOpen** in finally block only once out of for loop. So due to that when we try to perform open same file in below code
> smbJFile = share.openFile(outputPath, EnumSet.of(AccessMask.GENERIC_WRITE, AccessMask.GENERIC_READ), null, null,
						SMB2CreateDisposition.FILE_OVERWRITE_IF, null);

it throws STATUS_SHARING_VIOLATION(3221225539/3221225539)  exception on this line.

**Soultion:- I have added finally block in for loop for each & every file close operations. & it resolved above issue.**

But now my question is that as i found one cause of **STATUS_SHARING_VIOLATION(3221225539/3221225539): Create failed for** in my code but not the same case at all.

So can you please guide me what are the other different causes for **STATUS_SHARING_VIOLATION(3221225539/3221225539): Create failed for** exception.

**Comment by pepijnve** [comment]
Please refer to the first big table at https://docs.microsoft.com/en-us/windows/desktop/FileIO/creating-and-opening-files with first / second CreateFile calls with all the access and share mode permutations.

**Comment by harisingh-highq** [comment]
@pepijnve 
Thanks for sharing reference link.

But i mean to ask that there is only and only possibility of our code handling problem or it might be thrown same exception due to network fluctuation with file server in our live environment or might be other different cause?
So we can decide that it is critical error or we can ignore such cases?

**Comment by pepijnve** [comment]
Please read the documentation thoroughly. Any and all `STATUS_...` errors originate from the SMB server you're communicating with; smbj does not produce these itself, nor does the local operating system of the client.

In my opinion it's out of scope for the smbj project to document all the intricacies of the smb protocol, there is ample detailed documentation about all these things on MSDN.
If you get some particular error code, look it up in the list of error codes at https://msdn.microsoft.com/en-us/library/cc704588.aspx. Then determine if this is caused by a programming error on your end or not and if it's something you want to handle in your code or not. I don't think anyone else can make that decision in your place.

In the concrete case of `STATUS_SHARING_VIOLATION` the docs state:
> A file cannot be opened because the share access flags are incompatible.
i.e., the file has already been opened by you or some other SMB client with a set of flags that is incompatible with the flags you specified for this open request. The table I referenced above explains which combination of share access flags are (in)compatible. This is (as far as I know) the only cause of `STATUS_SHARING_VIOLATION`.

Can you safely ignore this? I can't tell you that. Depends on what you're trying to do in your application.

**Comment by harisingh-highq** [comment]
@pepijnve 
Thanks for the explanation.
I have monitored our live environment in details and found that it throws STATUS_SHARING_VIOLATION(3221225539/3221225539) exception in file upload but after some time (means in next try), it will get success to file upload of same file. So i think it is not caused by programming error but it might be due to network fluctuation or anything else going wrong.

If it is caused by programming error, then i can regenerate it locally easily and every time it should throw but as per analysis of our application live environment, it can be different cause.

**Comment by harisingh-highq** [comment]
> Sharing violation really only means that some other process has opened the file with a set of flags that are incompatible with the ones you've requested; nothing else.
yes. But i have reviewed all smbj code in my application and found same lines of code for write file as below:-
`smbJFile = share.openFile(outputPath, EnumSet.of(AccessMask.GENERIC_WRITE, AccessMask.GENERIC_READ), null, null,
						SMB2CreateDisposition.FILE_OVERWRITE_IF, null);`

But i am little bit confused about that it works fine in all operations but only sometimes it throws STATUS_SHARING_VIOLATION(3221225539/3221225539)  exception.

Do you have idea about any tool to monitor file open/access by other process?

**Comment by pepijnve** [comment]
In distributed systems things happen asynchronously. The SMB server, some other process on the server (e.g., a virus scanner, an indexing system, ...) or some other client might have the file open for a brief period. If you try to open the file during that time window, you'll get this error. That's just the nature of this type of issue.

The only authoritative source for open file information is the SMB server itself in this case. As far as I know there's nothing in the core SMB protocol to query this. There might be some RPC call you could use, not sure. Support may vary from server to server.
On the server side you typically have some tools to monitor or at least query this information. Tools are operating system dependent.

**Comment by harisingh-highq** [comment]
okay @pepijnve 
Thanks for your valuable time. I appreciate your help to explain this issue.

> You'll have to root cause this yourself.

I will try to look out it in more details.

**Comment by hierynomus** [comment]
@harisingh-highq I assume this has been answered to your satisfaction. If so, could you close this issue? Thanks!

**Comment by hierynomus** [comment]
@harisingh-highq Can this be closed?
